File: uplinks/kos/mod/KosDataParser.cs
// GonogoKosUplink: GPLv3. See GonogoKosUplink.csproj's header comment for the
// licence/linkage rationale.

using System;
using System.Collections.Generic;
using System.Globalization;
using System.Text.RegularExpressions;

namespace Gonogo.KosUplink
{
    /// <summary>
    /// Pure, KSP/kOS-free parser for the <c>[KOSDATA:&lt;id&gt;]k=v;k=v[/KOSDATA]</c>
    /// wire format kOS compute scripts emit via <c>PRINT</c>, the C# port of
    /// the app-side <c>packages/data/src/kos/kos-data-parser.ts</c>, moved
    /// mod-side per <c>kos-migration-spec.md</c> §4(b): in-process we capture
    /// the clean, un-wrapped source text at <c>ScreenBuffer.Print</c> (the
    /// Harmony postfix), so grid-reassembly and ANSI-splitting largely go
    /// away: but the marker/body grammar and the value coercion rules stay
    /// byte-identical to the TS so the client sees exactly the same values.
    ///
    /// <para>ANSI stripping is retained defensively (the snapshot-scrape
    /// fallback, spec §4.4, still feeds reassembled grid rows) and is a cheap
    /// no-op when the text carries no escape byte, exactly as the TS does.</para>
    /// </summary>
    public static class KosDataParser
    {
        /// <summary>Topic id used when a block omits the <c>:topic</c> suffix, mirrors the TS <c>DEFAULT_KOS_TOPIC</c>.</summary>
        public const string DefaultTopic = "default";

        private const char Esc = '\u001b';

        // Group 1 = optional topic id ("shipmap" in [KOSDATA:shipmap]); group
        // 2 = body. Topic charset is [\w-] so kebab-case ids don't clash with
        // the body's ; or =. Mirrors the TS BLOCK_RE exactly.
        private static readonly Regex BlockRe = new Regex(
            @"\[KOSDATA(?::([\w-]+))?\]([\s\S]*?)\[/KOSDATA\]",
            RegexOptions.Compiled);

        // CSI / OSC / bare 2-byte escapes, mirrors the TS ANSI_RE verbatim
        // ( = ESC,  = BEL).
        private static readonly Regex AnsiRe = new Regex(
            "\u001b\\[[0-?]*[ -/]*[@-~]|\u001b\\][^\u0007\u001b]*(?:\u0007|\u001b\\\\)|\u001b[@-Z\\\\-_?]",
            RegexOptions.Compiled);

        /// <summary>
        /// Strips ANSI control sequences. Cheap short-circuit when the text
        /// contains no ESC byte (the common plain-<c>PRINT</c> case), same as
        /// the TS <c>stripAnsi</c>.
        /// </summary>
        public static string StripAnsi(string text)
        {
            if (text == null) return "";
            if (text.IndexOf(Esc) == -1) return text;
            return AnsiRe.Replace(text, "");
        }

        /// <summary>
        /// Topic-aware parse. Returns the latest body per topic id seen in
        /// <paramref name="text"/>, keyed by topic; bare <c>[KOSDATA]</c>
        /// blocks key under <see cref="DefaultTopic"/>. When two blocks share
        /// a topic the later one wins (newer beats older). Returns an empty
        /// dictionary when no complete block is present (the C# analogue of
        /// the TS returning <c>null</c>: an empty result reads the same at
        /// every call site without a null-check).
        /// </summary>
        public static Dictionary<string, Dictionary<string, object>> ParseTopics(string text)
        {
            var result = new Dictionary<string, Dictionary<string, object>>(StringComparer.Ordinal);
            var clean = StripAnsi(text);
            foreach (Match m in BlockRe.Matches(clean))
            {
                var topic = m.Groups[1].Success ? m.Groups[1].Value : DefaultTopic;
                result[topic] = ParseBody(m.Groups[2].Value);
            }
            return result;
        }

        /// <summary>
        /// Parses one block body (<c>k=v;k=v</c>) into a field map. Blank keys
        /// and entries with no <c>=</c> are skipped, matching the TS
        /// <c>parseBody</c>.
        /// </summary>
        public static Dictionary<string, object> ParseBody(string body)
        {
            var outMap = new Dictionary<string, object>(StringComparer.Ordinal);
            if (body == null) return outMap;
            foreach (var raw in body.Split(';'))
            {
                var eq = raw.IndexOf('=');
                if (eq == -1) continue;
                var key = raw.Substring(0, eq).Trim();
                if (key.Length == 0) continue;
                var value = raw.Substring(eq + 1).Trim();
                outMap[key] = Coerce(value);
            }
            return outMap;
        }

        // Must accept "-1.5", "3e-2", "0"; rejects "NaN" (ambiguous, surfaced
        // as a string so the widget decides). Mirrors the TS coerce() and its
        // number regex exactly. A JSON value (parts=<json>) fails the numeric
        // test and passes through as a string, which the client JSON.parses.
        private static readonly Regex NumberRe = new Regex(
            @"^-?(?:\d+\.?\d*|\.\d+)(?:[eE][+-]?\d+)?$", RegexOptions.Compiled);

        /// <summary>
        /// Coerces one raw value string to <see cref="bool"/> / <see cref="double"/>
        /// / <see cref="string"/>: the exact rules of the TS <c>coerce</c>.
        /// </summary>
        public static object Coerce(string value)
        {
            if (value == "true") return true;
            if (value == "false") return false;
            if (value.Length != 0 && NumberRe.IsMatch(value))
            {
                return double.Parse(value, NumberStyles.Float, CultureInfo.InvariantCulture);
            }
            return value;
        }
    }
}

File: uplinks/kos/client/src/shared/kos-data-parser.ts
/**
 * Parser for the `[KOSDATA] k=v;k=v [/KOSDATA]` wire format that kOS widget
 * scripts MUST emit on stdout. Pure function, no I/O.
 *
 * The input is any chunk of kOS terminal output, the parser locates the
 * marker pair, parses the key/value body, and returns an object. Text
 * outside the markers (REPL prompt, RUN echo, stray PRINTs) is ignored.
 *
 * If the chunk contains multiple `[KOSDATA] ... [/KOSDATA]` blocks, the last
 * one wins. That matches the intended contract: scripts emit exactly one,
 * so if we see more than one the later one is always newer.
 *
 * Topics: blocks may optionally carry a topic id like `[KOSDATA:shipmap]...
 * [/KOSDATA]`, used by the centralised kOS compute fanout to multiplex
 * several feeds on a single CPU's print stream. `parseKosData` ignores
 * the topic and returns the body of the last block found (matching the
 * legacy single-block contract); `parseKosDataTopics` returns one entry
 * per topic, with bare `[KOSDATA]` keyed under `default`.
 */

import { createPerfBudget } from "@ksp-gonogo/sitrep-sdk";

/**
 * Soft cap on parser invocations. The proxy emits one PTY chunk per
 * line of kOS output; with multiple kOS widgets polling every few
 * seconds, expect ~10–30/sec under normal load. Threshold at 200/sec
 * catches infinite-loop scripts or runaway kOS PRINTs that would flood
 * the parse pipeline.
 */
const KOS_PARSE_BUDGET = createPerfBudget({
  name: "kos-data-parser.parseKosData calls/sec",
  threshold: 200,
  windowMs: 1000,
  unit: "calls",
});

export type KosDataValue = number | boolean | string;
export type KosData = Record<string, KosDataValue>;

/** Runtime-resolved arg value passed to a kOS compute data source. */
export type KosScriptArg = number | string | boolean;

/**
 * Group 1 = optional topic id (`shipmap` in `[KOSDATA:shipmap]`); undefined
 * for bare `[KOSDATA]`. Group 2 = body. Topic charset is `[\w-]` so script
 * authors can use kebab-case ids without clashing with `;` or `=` in the body.
 */
const BLOCK_RE = /\[KOSDATA(?::([\w-]+))?\]([\s\S]*?)\[\/KOSDATA\]/g;

/** Topic id used when a block omits the `:topic` suffix. */
export const DEFAULT_KOS_TOPIC = "default";

/**
 * Strip ANSI control sequences. kOS's GUI repaint emits screen contents
 * with a cursor-position escape (`ESC [ row ; col H`) injected at every
 * terminal line wrap, which can split our `[KOSDATA]` marker, observed
 * in the wild as `[/KOSDA<ESC[22;1H>TA]`. Stripping these BEFORE the
 * marker scan makes the parser robust to wrapping across PTY rows.
 *
 * Sequences covered:
 *   - CSI: `ESC [ params final-byte`
 *   - OSC: `ESC ] ... BEL` (title-set, etc.)
 *   - Bare 2-byte escapes: `ESC <letter>` or `ESC ?` etc.
 */
const ANSI_RE =
  // biome-ignore lint/suspicious/noControlCharactersInRegex: stripping ANSI is the whole point
  /\x1b\[[0-?]*[ -/]*[@-~]|\x1b\][^\x07\x1b]*(?:\x07|\x1b\\)|\x1b[@-Z\\-_?]/g;

export function stripAnsi(text: string): string {
  // Most plain-PRINT chunks have no escape character; the indexOf check
  // is roughly an order of magnitude cheaper than running the regex
  // `replace` blindly. Worth doing because parseKosData is called per
  // PTY chunk during active kOS widget polling.
  if (text.indexOf("\x1b") === -1) return text;
  return text.replace(ANSI_RE, "");
}

/**
 * Returns the parsed key/value object from the last `[KOSDATA]` block in
 * `text`, or null if no complete block is present.
 */
export function parseKosData(text: string): KosData | null {
  KOS_PARSE_BUDGET.record();
  const clean = stripAnsi(text);
  let lastBody: string | null = null;
  // Reset lastIndex each call: BLOCK_RE is module-scoped.
  BLOCK_RE.lastIndex = 0;
  let match = BLOCK_RE.exec(clean);
  while (match !== null) {
    lastBody = match[2];
    match = BLOCK_RE.exec(clean);
  }
  if (lastBody === null) return null;
  return parseBody(lastBody);
}

/**
 * Topic-aware parse. Returns the latest body per topic id seen in `text`,
 * keyed by topic. Bare `[KOSDATA]` blocks are keyed under `DEFAULT_KOS_TOPIC`.
 *
 * Returns `null` if no complete block is present at all (parity with
 * `parseKosData`). When two blocks share a topic id, the later one wins,
 * same "newer block beats older" rule as the single-block parser.
 */
export function parseKosDataTopics(text: string): Map<string, KosData> | null {
  KOS_PARSE_BUDGET.record();
  const clean = stripAnsi(text);
  const result = new Map<string, KosData>();
  BLOCK_RE.lastIndex = 0;
  let match = BLOCK_RE.exec(clean);
  while (match !== null) {
    const topic = match[1] ?? DEFAULT_KOS_TOPIC;
    result.set(topic, parseBody(match[2]));
    match = BLOCK_RE.exec(clean);
  }
  if (result.size === 0) return null;
  return result;
}

function parseBody(body: string): KosData {
  const out: KosData = {};
  for (const raw of body.split(";")) {
    const eq = raw.indexOf("=");
    if (eq === -1) continue;
    const key = raw.slice(0, eq).trim();
    if (key === "") continue;
    const value = raw.slice(eq + 1).trim();
    out[key] = coerce(value);
  }
  return out;
}

function coerce(value: string): KosDataValue {
  if (value === "true") return true;
  if (value === "false") return false;
  // Must accept things like "-1.5", "3e-2", "0". Rejects "NaN" (ambiguous,
  // we'd rather surface it as a string so the widget can decide).
  if (value !== "" && /^-?(?:\d+\.?\d*|\.\d+)(?:[eE][+-]?\d+)?$/.test(value)) {
    return Number(value);
  }
  return value;
}
